### ISSUE TITLE
[Practice App] Art Item Related Features

### ISSUE THREAD

**Issue by canatakan** [bounswe/bounswe2022group7#173]
In [Meeting #12](https://github.com/bounswe/bounswe2022group7/wiki/Meeting-Notes-%2312), we assigned features to team members. I was assigned to handle the art item related features.

# Requirements

- Users shall be able to view an art item in the platform
- Users shall be able to view dominant colors in an art item
- Artists shall be able to create an art item in the platform
- The art item shall include a name, a description, and an image

# Use Case Diagram
<img width="728" alt="art_item_use_case" src="https://user-images.githubusercontent.com/56476673/168831181-62dca0f3-cbb5-4751-ae2c-9741da922b56.png">

# Class Diagram
<img width="1032" alt="art_item_class" src="https://user-images.githubusercontent.com/56476673/168836003-99fb9341-77ff-4737-9085-53c93328afa4.png">

# Sequence Diagrams

## View Art Item

![art_item_seq_view](https://user-images.githubusercontent.com/56476673/168857026-68660908-f550-46c3-b1ba-498ef90d3a30.png)

## Create Art Item

![art_item_seq_create](https://user-images.githubusercontent.com/56476673/168857199-f3bb1a47-2013-4ff2-b2e8-e1c8baab1f40.png)

# Feature Implementations
- [x] **Create a new branch based on the [practice_app](https://github.com/bounswe/bounswe2022group7/tree/practice_app) branch**
    According to the naming convention that we discussed, the name of the branch shall be `practice-app/feature/art-items`.
- [x] Create an endpoint for creating art items
- [x] Create an endpoint for viewing a particular art item
- [x] Connect the frontend templates to the api endpoints
- [x] Update the sidebar in the [base.html](https://github.com/bounswe/bounswe2022group7/blob/practice_app/practice-app/website/templates/base.html) by adding a link to the art item creation page
- [x] Create a PR and add @AlicanM as a reviewer.

**Deadline: 18/05/2022 23:59**

# Testing & Documentation
Once all the work is completed, I need to add:
- [x] #183
- [x] #200
- [x] Create .yml files for Swagger API documentation (See #192 for tracking)

These subtasks mentioned above shall be converted to separate issues with their deadlines included.

# Conclusion

Once all the tasks are completed, a PR will be created and @AlicanM will be added as a reviewer.

**Comment by canatakan** [comment]
* I decided to use an external API that extracts the dominant colors from an image URL. The API provider that I decided to use is [Imagga](https://imagga.com/).

* Thanks to this API's `color` endpoint, it shall be possible to obtain most dominant colors in the image URL that the user provided while creating the art item.

* The API documentation can be found [here](https://docs.imagga.com/).

**Comment by canatakan** [comment]
Also, I think the most suitable way for making requests to an external API is to use the `requests` module. Therefore, the `requirements.txt` file might need a change.

**Comment by canatakan** [comment]
### 1. Branch is created, and the files are added.

* You can check [the initial commit](https://github.com/bounswe/bounswe2022group7/commit/dab40ffe61299208cfe8670712f61847c4e44bc8).
* The branch named `practice-app/feature/art-items` is created.
* I changed these files:
  *  [\_\_init.py\_\_](https://github.com/bounswe/bounswe2022group7/commit/dab40ffe61299208cfe8670712f61847c4e44bc8#diff-1bee5bb9edb3a1e35a9d9a69ecff98e8497e8ad1f2428ae0d60d548641883af6)
  * [art_item.py](https://github.com/bounswe/bounswe2022group7/commit/dab40ffe61299208cfe8670712f61847c4e44bc8#diff-05fec614af19861ed335825727985d2076cc40345a81669c4217e40104def7a4)
  * [settings.py](https://github.com/bounswe/bounswe2022group7/commit/dab40ffe61299208cfe8670712f61847c4e44bc8#diff-8da87a37f8b2442721c62cdeec2dfb1a65566ef95af29d2e2b9fd737cc146276)
  * [create_art_item.html](https://github.com/bounswe/bounswe2022group7/commit/dab40ffe61299208cfe8670712f61847c4e44bc8#diff-41d991d481447a5f9f9f8af3208b1bdd5d6a3312ca2fb794c243ca9c05b89ea1)
  * [view_art_item.html](https://github.com/bounswe/bounswe2022group7/commit/dab40ffe61299208cfe8670712f61847c4e44bc8#diff-d4431cfad971fcd904ee7d41399295738541d5b1134daf5ea949041dd02d5117)
  * [views.py](https://github.com/bounswe/bounswe2022group7/commit/dab40ffe61299208cfe8670712f61847c4e44bc8#diff-0625e365e84831d1dbb019929602463f826ccdec2331709db07f4a20143ad7e9)

### 2. The endpoints are created.

* **The endpoint for creating an art item is added.**
    The route for this endpoint can be viewed in [art_item.py](https://github.com/bounswe/bounswe2022group7/blob/practice-app/feature/art_items/practice-app/website/api/art_item.py) as `@art_item.route("/art_item", methods=["POST"])`.

* **The endpoint for viewing an art item is added.**
    The route for this endpoint can be viewed in [art_item.py](https://github.com/bounswe/bounswe2022group7/blob/practice-app/feature/art_items/practice-app/website/api/art_item.py) as `@art_item.route("/art_item/<art_item_id>", methods=["GET"])`.

* **The API keys for the external API call are added to [settings.py](https://github.com/bounswe/bounswe2022group7/blob/practice-app/feature/art_items/practice-app/website/settings.py).**
   You can see them as below:
    ```
    COLOR_FINDER_API_KEY =  os.environ.get("COLOR_FINDER_API_KEY")
    COLOR_FINDER_API_SECRET = os.environ.get("COLOR_FINDER_API_SECRET")
    ``` 

* The blueprints for the routes are registered in [\_\_init.py\_\_](https://github.com/bounswe/bounswe2022group7/blob/practice-app/feature/art_items/practice-app/website/__init__.py).

### 3. Front-end templates are connected to the API endpoints.

* You can check [views.py](https://github.com/bounswe/bounswe2022group7/blob/practice-app/feature/art_items/practice-app/website/views.py), [create_art_item.html](https://github.com/bounswe/bounswe2022group7/blob/practice-app/feature/art_items/practice-app/website/templates/create_art_item.html), and [view_art_item.html](https://github.com/bounswe/bounswe2022group7/blob/practice-app/feature/art_items/practice-app/website/templates/view_art_item.html) files to examine the connection between front-end templates and APIs.

**Comment by canatakan** [comment]
> Also, I think the most suitable way for making requests to an external API is to use the `requests` module. Therefore, the `requirements.txt` file might need a change.

Noticed that `requests` module is already obtained with the current `requirements.txt` file. No changes are made for this, please ignore.

**Comment by canatakan** [comment]
The sidebar in the [base.html](https://github.com/bounswe/bounswe2022group7/blob/practice-app/feature/art_items/practice-app/website/templates/base.html) file is updated with the link to the art item creation page. You can check the relevant commit, [here](https://github.com/bounswe/bounswe2022group7/commit/b79616461718e7dd053feb6c85b541d03dc35883).

**Comment by canatakan** [comment]
* All files required for the Swagger API documentation are added. Relevant commit is available [here](https://github.com/bounswe/bounswe2022group7/commit/552de245906a0d3c785b9dc877a0068879ede68e). 

* You can find the newly created `.yml` files below:
    [art_item_GET.yml](https://github.com/bounswe/bounswe2022group7/blob/practice-app/feature/art-items/practice-app/website/api/doc/art_item_GET.yml)
    [art_item_POST.yml](https://github.com/bounswe/bounswe2022group7/blob/practice-app/feature/art-items/practice-app/website/api/doc/art_item_POST.yml)

**Comment by canatakan** [comment]
* All tasks except unit testing are done. I have just added the API documentation to `practice-app/feature/art-items` branch, too. 

* For other members to benefit from the API documentation files, I have created a pull request (#193) for merging `practice-app/feature/art-items` branch to `practice_app` branch, before the completion of unit tests. 

* Unit tests shall be merged to the `practice_app` branch after completion.

* The pull request is now ready for your review @AlicanM.

**Comment by canatakan** [comment]
Hi @AlicanM,

Since the unit tests are now also completed, I have just sent another pull request. You can find the pull request in #206. The review may not take too long as only the unit tests are added.

**Comment by canatakan** [comment]
Also, after this achievement, all subtasks for implementing Art Item related features are done. Hence, marking the issue as "Pending Review".

**Comment by AlicanM** [comment]
I merged the pull request #206, thanks for your work.
